test/telepilot_test.go

```diff
@@ -0,0 +1,173 @@
+	if _, err := os.Stat(path.Join(certDir, name+".pem")); err != nil {
+		t.Skip("Missing cert, skipping. Make sure to run `make mtls` first or invoke tests with `make test`.")
+	}
+	cfg, err := tlsconfig.LoadTLSConfig(
+		path.Join(certDir, name+".pem"),
+		path.Join(certDir, name+"-key.pem"),
+		path.Join(certDir, "ca.pem"),
+		name != "server",
+	)
+	noError(t, err, "Load certs for "+name)
+	return cfg
+}
+
+func TestStartStop(t *testing.T) {
+	t.Parallel()
+
+	// Load certs for the server and a couple of clients.
+	serverTLSConfig := loadTLSConfig(t, "server")
+	aliceTLSConfig := loadTLSConfig(t, "client-alice")
+	bobTLSConfig := loadTLSConfig(t, "client-bob")
+
+	// Create a server.
+	srv := apiserver.NewServer(serverTLSConfig)
+
+	// Listen on a random port.
+	lis, err := net.Listen("tcp", "127.0.0.1:0")
+	noError(t, err, "Listen")
+	t.Cleanup(func() { _ = lis.Close() }) // No strictly needed, but just to make sure. Called by the server Close().
+
+	// Start the server.
+	doneCh := make(chan struct{})
+	go func() { defer close(doneCh); noError(t, srv.Serve(lis), "Serve") }()
+	t.Cleanup(func() { noError(t, srv.Close(), "Closing server."); <-doneCh }) // Wait for the server to be fully closed.
+
+	// Create clients.
+	aliceClient, err := apiclient.NewClient(aliceTLSConfig, lis.Addr().String())
+	noError(t, err, "NewClient for Alice")
+	t.Cleanup(func() { noError(t, aliceClient.Close(), "Closing Alice's client.") })
+
+	bobClient, err := apiclient.NewClient(bobTLSConfig, lis.Addr().String())
+	noError(t, err, "NewClient for Bob")
+	t.Cleanup(func() { noError(t, bobClient.Close(), "Closing Bob's client.") })
+
+	ctx := context.Background()
+
+	// Create a Job.
+	jobID, err := aliceClient.StartJob(ctx, "foo", nil)
+	noError(t, err, "Alice start job.")
+
+	// Stop the Job from the same user.
+	t.Run("happy alice", func(t *testing.T) {
+		t.Parallel()
+		st, ok := status.FromError(aliceClient.StopJob(ctx, jobID))
+		assert(t, true, ok, "extract grpc status from error")
+
+		// TODO: Change this to noError once implemented.
+		assert(t, codes.Unimplemented, st.Code(), "invalid grpc status code")
+	})
+
+	// Attempt to stop the job from a different user.
+	t.Run("sad bob", func(t *testing.T) {
+		t.Parallel()
+
+		st, ok := status.FromError(bobClient.StopJob(ctx, jobID))
+		assert(t, true, ok, "extract grpc status from error")
+
+		assert(t, codes.PermissionDenied, st.Code(), "invalid grpc status code")
+
+		// Attempt to stop a non-exiting job.
+		st, ok = status.FromError(bobClient.StopJob(ctx, uuid.New().String()))
+		assert(t, true, ok, "extract grpc status from error")
+
+		assert(t, codes.PermissionDenied, st.Code(), "invalid grpc status code")
+	})
+}
+
+func TestUnauthenticatedUser(t *testing.T) {
+	t.Parallel()
+
+	// Load certs for the server and a couple of clients.
+	serverTLSConfig := loadTLSConfig(t, "server")
+	aliceTLSConfig := loadTLSConfig(t, "client-alice")
```

Comment:
Could we use certs from an unknown CA, so the test is more realistic?

Reply:
Added a TODO. Agreed, will do that in a future PR.